translate/compiler.go

```diff
@@ -16,15 +16,130 @@
 
 package translate
 
-import "github.com/DE-labtory/koa/ast"
+import (
+	"github.com/DE-labtory/koa/ast"
+	"github.com/DE-labtory/koa/encoding"
+	"github.com/DE-labtory/koa/opcode"
+	"github.com/pkg/errors"
+)
 
 type Compiler struct {
+	// binary is the byte code which is compiled.
+	binary []byte
+
+	debug bool
+}
+
+func NewCompiler(debug bool) *Compiler {
+	return &Compiler{
+		binary: make([]byte, 0),
+		debug:  debug,
+	}
+}
+
+// Compile() compiles statements in program.
+// Statements would be compiled to byte code.
+// TODO: implement w/ test cases :-)
+func (c *Compiler) Compile(program ast.Program) error {
+	for _, s := range program.Statements {
+		if err := c.compileNode(s); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// emit() generates a byte code with operator and operands.
+// Then, adds the byte code to binary in compiler
+// and returns the position of this instruction.
+// TODO: implement w/ test cases :-)
+func (c *Compiler) emit(operator opcode.Type, operand ...[]byte) int {
+	return 0
 }
 
-func NewCompiler() *Compiler {
+// compileNode() compiles a node in statement.
+// This function will be executed recursively.
+// TODO: implement w/ test cases :-)
+func (c *Compiler) compileNode(node ast.Node) error {
+	// Nodes are many kinds.
+	switch node := node.(type) {
+	case *ast.Identifier:
+		if err := c.compileIdentifier(*node); err != nil {
+			return err
+		}
+
+	case *ast.AssignStatement:
+		if err := c.compileAssignStatement(*node); err != nil {
+			return err
+		}
+
+	case *ast.StringLiteral:
+		if err := c.compileString(*node); err != nil {
+			return err
+		}
+
+	case *ast.IntegerLiteral:
+		if err := c.compileInteger(*node); err != nil {
+			return err
+		}
+
+	case *ast.BooleanLiteral:
+		if err := c.compileBoolean(*node); err != nil {
+			return err
+		}
+
+	case *ast.PrefixExpression:
+		if err := c.compilePrefixExpression(*node); err != nil {
+			return err
+		}
+
+	default:
+		return errors.New("compileNode() error - " + node.String() + " could not compiled")
+	}
+
 	return nil
 }
 
-func (c *Compiler) Compile(program ast.Program) {
+// TODO: implement w/ test cases :-)
+func (c *Compiler) compileIdentifier(node ast.Identifier) error {
```

Comment:
Are these functions ethereum style?

Reply:
Yes they are, but I mixed the go ethereum style and the monkey style.


Go ethereum seperates only element, number, eof, etc. (go ether reads line by line.)
Monkey seperates string, integer, boolean, prefix, infix, etc. But monkey doesn’t make functions. It is only written by switch/case statements.

So, I seperates them like the monkey, but functions style came from go ethereum.